pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1021,12 +1024,34 @@ public void connectionFailed(PulsarClientException exception) {
                 if (ignoreException && exception != null) {
                     log.debug("Exception ignored in closing consumer", exception);
                 }
-                cleanupAtClose(closeFuture, ignoreException ? null : exception);
+                cleanupAtClose(closeConsumerFuture, ignoreException ? null : exception);
                 return null;
             });
         }
+        closeFutures.add(closeConsumerFuture);
+        return CompletableFuture.allOf(closeFutures.toArray(new CompletableFuture[0]));
+    }
 
-        return closeFuture;
+    private List<CompletableFuture<Void>> closeDeadAndRetryProducer() {
+        final List<CompletableFuture<Void>> closeFutures = new ArrayList<>();
+        if (this.deadLetterProducer != null) {
+            try {
+                deadLetterProducer.whenComplete((r, t) -> {
```

Comment:
deadLetterProducer.closeAsync()?

Reply:
deadLetterProducer is future ,not producer